### ISSUE TITLE
`DistributedDataParallel` with `torchvision.models.resnet152` produces inconsistent gradients

### ISSUE THREAD

**Issue by andwgu** [pytorch/pytorch#62062]
## 🐛 Bug

Using `torchvision.models.resnet152` wrapped with `DistributedDataParallel` produces inconsistent gradients when run on the same input.

## To Reproduce

```
import copy
import os
import torch
import torch.distributed as dist
import torch.multiprocessing as mp
import torch.nn as nn
import torchvision
from torch.nn.parallel import DistributedDataParallel as DDP

BACKEND = "nccl"
WORLD_SIZE = 2
BATCH_SIZE = 32

def worker(rank, world_size):
    torch.manual_seed(0)
    torch.cuda.manual_seed(0)
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '29500'
    dist.init_process_group(BACKEND, rank=rank, world_size=world_size)
    torch.cuda.set_device(rank)

    model = torchvision.models.resnet152(pretrained=False)
    ddp_model1 = DDP(model.to(rank), device_ids=[rank], gradient_as_bucket_view=True)
    ddp_model2 = copy.deepcopy(ddp_model1)

    input = torch.randn(BATCH_SIZE, 3, 224, 224).to(rank)
    labels = torch.randn(BATCH_SIZE, 1000).to(rank)
    loss_fn = nn.MSELoss()

    grads1 = []
    output1 = ddp_model1(input)
    loss1 = loss_fn(output1, labels)
    loss1.backward()
    torch.cuda.synchronize()
    for p in ddp_model1.parameters():
        grads1.append(p.grad.clone())

    grads2 = []
    output2 = ddp_model2(input)
    loss2 = loss_fn(output2, labels)
    loss2.backward()
    torch.cuda.synchronize()
    for p in ddp_model2.parameters():
        grads2.append(p.grad.clone())

    if rank == 0:
        assert torch.allclose(output1, output2)
        for i, (g1, g2) in enumerate(zip(grads1, grads2)):
            if not torch.allclose(g1, g2):
                diff = torch.max(abs(g1 - g2)).item()
                if diff > 1e-5:
                    print(f"Significant mismatch on grad {i}: {diff:.8f}")

def main():
    world_size = WORLD_SIZE
    mp.spawn(worker,
             args=(world_size,),
             nprocs=world_size,
             join=True)

if __name__ == "__main__":
    main()
```
The program creates a `resnet152` DDP model and deep copies it. Then, it runs the same input through both models. The outputs match, but the gradients produced do not (with some variation from run to run):
```
Significant mismatch on grad 0: 0.00008440
Significant mismatch on grad 1: 0.00001209
Significant mismatch on grad 6: 0.00003475
Significant mismatch on grad 18: 0.00001973
Significant mismatch on grad 27: 0.00001633
```

## Expected behavior

The program should run to termination without any output.

## Environment

 - PyTorch Version (e.g., 1.0): 1.10
 - OS (e.g., Linux): Ubuntu 18.04.5 LTS (x86_64)
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Build command you used (if compiling from source): `python setup.py develop`
 - Python version: 3.8 (64-bit runtime)
 - CUDA/cuDNN version: 11.1.105
 - GPU models and configuration: 2 Tesla V100
 - Any other relevant information: 

## Additional context

This may just be due to numerical issues (e.g. floating point arithmetic not being associative) and not a real problem with `DistributedDataParallel`. Nonetheless, I wanted to file the issue since it made checking for parity between two approaches difficult when using `resnet152` as the model. 

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

**Comment by rohan-varma** [comment]
In addition to setting seed, also try wrapping test with this:
```
with torch.backends.cudnn.flags(enabled=True, deterministic=True, benchmark=False):
    # your test
```

to enable determinism for cudnn operators

**Comment by andwgu** [comment]
> In addition to setting seed, also try wrapping test with this:
> 
> ```
> with torch.backends.cudnn.flags(enabled=True, deterministic=True, benchmark=False):
>     # your test
> ```
> 
> to enable determinism for cudnn operators

This fixed it! Thanks @rohan-varma :)
